---
title: "UCSCXenaTools: Download Public Cancer Genomic Data from UCSC Xena Hubs"
author: "Shixiang Wang \\

        ShanghaiTech. University"
date: "`r Sys.Date()`"

output:
  prettydoc::html_pretty:
    toc: true
    theme: cayman
    highlight: github
  pdf_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{download publical data in batch quickly and easily}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

**Current Version: 0.2.4**

`UCSCXenaTools` is a R package download and explore data from **UCSC Xena data hubs**, which are

> A collection of UCSC-hosted public databases such as TCGA, ICGC, TARGET, GTEx, CCLE, and others. Databases are normalized so they can be combined, linked, filtered, explored and downloaded.
>
> -- [UCSC Xena](https://xena.ucsc.edu/)


## Installation

Install stable release from CRAN with:

```{r, eval=FALSE}
install.packages("UCSCXenaTools")
```


You can also install devel version of UCSCXenaTools from github with:

```{r gh-installation, eval = FALSE}
# install.packages("devtools")
devtools::install_github("ShixiangWang/UCSCXenaTools", build_vignettes = TRUE)
```

Read this vignettes.

```{r, eval=FALSE}
browseVignettes("UCSCXenaTools")
# or
??UCSCXenaTools
```


## Data Hub List

All datasets are available at <https://xenabrowser.net/datapages/>.

Currently, `UCSCXenaTools` support all 7 data hubs of UCSC Xena.

* [UCSC Public Hub](https://xenabrowser.net/datapages/?host=https%3A%2F%2Fucscpublic.xenahubs.net): <https://ucscpublic.xenahubs.net>
* [TCGA Hub](https://xenabrowser.net/datapages/?host=https%3A%2F%2Ftcga.xenahubs.net): <https://tcga.xenahubs.net>
* [GDC Xena Hub](https://xenabrowser.net/datapages/?host=https%3A%2F%2Fgdc.xenahubs.net): <https://gdc.xenahubs.net>
* [ICGC Xena Hub](https://xenabrowser.net/datapages/?host=https%3A%2F%2Ficgc.xenahubs.net): <https://icgc.xenahubs.net>
* [Pan-Cancer Atlas Hub](https://xenabrowser.net/datapages/?host=https%3A%2F%2Fpancanatlas.xenahubs.net): <https://pancanatlas.xenahubs.net>
* [GA4GH (TOIL) Hub](https://xenabrowser.net/datapages/?host=https%3A%2F%2Ftoil.xenahubs.net): <https://toil.xenahubs.net>
* [Treehouse Hub](https://xenabrowser.net/datapages/?host=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu): <https://xena.treehouse.gi.ucsc.edu>

If the `API` changed, please remind me by email to <w_shixiang@163.com> or open an issue on [GitHub](https://github.com/ShixiangWang/UCSCXenaTools/issues).

## Usage

Download UCSC Xena Datasets and load them into R by `UCSCXenaTools` is a workflow in `generate`, `filter`, `query`, `download` and `prepare` 5 steps, which are implemented as `XenaGenerate`, `XenaFilter`, `XenaQuery`, `XenaDownload` and `XenaPrepare`, respectively. They are very clear and easy to use and combine with other packages like `dplyr`.

The following use clinical data download of LUNG, LUAD, LUSC from TCGA (hg19 version) as an example.


### XenaData data.frame

Begin from version `0.2.0`, `UCSCXenaTools` use a `data.frame` object (built in package) `XenaData` to generate an instance of `XenaHub` class, which communicate with API of UCSC Xena Data Hubs.

You can load `XenaData` after loading `UCSCXenaTools` into R.

```{r}
library(UCSCXenaTools)
data(XenaData)

head(XenaData)
```



### Generate a XenaHub object

This can be implemented by `XenaGenerate` function, which generate `XenaHub` object from `XenaData` data frame.

```{r}
XenaGenerate()
```

We can set `subset` argument to narrow datasets.

```{r}
XenaGenerate(subset = XenaHostNames=="TCGA")
```

>You can use `XenaHub()` to generate a `XenaHub` object for API communication, but it is not recommended. 

It's possible to explore `hosts()`, `cohorts()` and `datasets()`.

```{r}
xe = XenaGenerate(subset = XenaHostNames=="TCGA")
# get hosts
hosts(xe)
# get cohorts
head(cohorts(xe))
# get datasets
head(datasets(xe))
```

Pipe operator `%>%` can also be used here.


```
> library(tidyverse)
> XenaData %>% filter(XenaHostNames == "TCGA", grepl("BRCA", XenaCohorts), grepl("Path", XenaDatasets)) %>% XenaGenerate()
class: XenaHub 
hosts():
  https://tcga.xenahubs.net
cohorts() (1 total):
  TCGA Breast Cancer (BRCA)
datasets() (4 total):
  TCGA.BRCA.sampleMap/Pathway_Paradigm_mRNA_And_Copy_Number
  TCGA.BRCA.sampleMap/Pathway_Paradigm_RNASeq
  TCGA.BRCA.sampleMap/Pathway_Paradigm_RNASeq_And_Copy_Number
  TCGA.BRCA.sampleMap/Pathway_Paradigm_mRNA
```

### Filter 

There are too many datasets, we filter them by `XenaFilter` function.

Regular expression can be used to filter XenaHub object to what we want.


```{r}
(XenaFilter(xe, filterDatasets = "clinical") -> xe2)
```


Then select `LUAD`, `LUSC` and `LUNG` 3 datasets.

```{r}
XenaFilter(xe2, filterDatasets = "LUAD|LUSC|LUNG") -> xe2
```

Pipe can be used here.

```{r}
suppressMessages(require(dplyr))

xe %>% 
    XenaFilter(filterDatasets = "clinical") %>% 
    XenaFilter(filterDatasets = "luad|lusc|lung")
```

### Query

Create a query before download data

```{r}
xe2_query = XenaQuery(xe2)
xe2_query
```

### Download

Default, data will be downloaded to system temp directory. You can specify the path.

If the data exists, command will not run to download them, but you can force it by `force` option.

```{r}
xe2_download = XenaDownload(xe2_query)
## not run
#xe2_download = XenaDownload(xe2_query, destdir = "E:/Github/XenaData/test/")
```

> Note fileNames transfromed from datasets name and / chracter all changed to __ character.

### Prepare

There are 4 ways to prepare data to R.

```
# way1:  directory
cli1 = XenaPrepare("E:/Github/XenaData/test/")
names(cli1)
## [1] "TCGA.LUAD.sampleMap__LUAD_clinicalMatrix.gz"
## [2] "TCGA.LUNG.sampleMap__LUNG_clinicalMatrix.gz"
## [3] "TCGA.LUSC.sampleMap__LUSC_clinicalMatrix.gz"
```

```
# way2: local files
cli2 = XenaPrepare("E:/Github/XenaData/test/TCGA.LUAD.sampleMap__LUAD_clinicalMatrix.gz")
class(cli2)
## [1] "tbl_df"     "tbl"        "data.frame"

cli2 = XenaPrepare(c("E:/Github/XenaData/test/TCGA.LUAD.sampleMap__LUAD_clinicalMatrix.gz",
                     "E:/Github/XenaData/test/TCGA.LUNG.sampleMap__LUNG_clinicalMatrix.gz"))
class(cli2)
## [1] "list"
names(cli2)
## [1] "TCGA.LUAD.sampleMap__LUAD_clinicalMatrix.gz"
## [2] "TCGA.LUNG.sampleMap__LUNG_clinicalMatrix.gz"
```

```
# way3: urls
cli3 = XenaPrepare(xe2_download$url[1:2])
names(cli3)
## [1] "LUSC_clinicalMatrix.gz" "LUNG_clinicalMatrix.gz"
```

```{r}
# way4: xenadownload object
cli4 = XenaPrepare(xe2_download)
names(cli4)
```


## TCGA Common Data Easy Download

### getTCGAdata

`getTCGAdata` provide a quite easy download way for TCGA datasets, user can specify multiple options to select which data and corresponding file type want to download. Default this function will return a list include `XenaHub` object and selective datasets information. Once you are sure the datasets is exactly what you want, `download` can be set to `TRUE` to download the data.

Check arguments of `getTCGAdata`:

```{r}
args(getTCGAdata)

# or run
# ??getTCGAdata to read documentation
```

Select one or more projects, default will select only clinical datasets:

```{r}
getTCGAdata(c("UVM", "LUAD"))

tcga_data = getTCGAdata(c("UVM", "LUAD"))

# only return XenaHub object
tcga_data$Xena

# only return datasets information
tcga_data$DataInfo
```

Set `download=TRUE` to download data, default data will be downloaded to system temp directory (you can specify the path with `destdir` option):

```{r}
# only download clinical data
getTCGAdata(c("UVM", "LUAD"), download = TRUE)
```

#### Support Data Type and Options

* clinical information: `clinical`
* mRNA Sequencing: `mRNASeq`
* mRNA microarray: `mRNAArray`
* miRNA Sequencing: `miRNASeq`
* exon Sequencing: `exonRNASeq`
* RPPA array: `RPPAArray`
* DNA Methylation: `Methylation`
* Gene mutation: `GeneMutation`
* Somatic mutation: `SomaticMutation`
* Gistic2 Copy Number: `GisticCopyNumber`
* Copy Number Segment: `CopyNumberSegment`

> other data type supported by Xena cannot download use this function. Please refer to `downloadTCGA` function or `XenaGenerate` function.


**NOTE**: Sequencing data are all based on  Illumina Hiseq platform, other platform (Illumina GA) data supported by Xena cannot download using this function. This is for building consistent data download flow. Mutation use `broad automated` version (except `PANCAN` use `MC3 Public Version`). If you wan to download other datasets, please refer to `downloadTCGA` function or `XenaGenerate` function.


### download any TCGA data by datatypes and filetypes 

`downloadTCGA` function can used to download any TCGA data supported by Xena, but in a way different from `getTCGAdata` function.

```{r, eval=FALSE}
# download RNASeq data (use UVM as an example)
downloadTCGA(project = "UVM",
                 data_type = "Gene Expression RNASeq",
                  file_type = "IlluminaHiSeq RNASeqV2")
```


See the arguments:

```{r}
args(downloadTCGA)
```

Except `destdir` option, you only need to select three arguments for downloading data. Even throught the number is far less than `getTCGAdata`, it is more complex than the latter.

Before you download data, you need spare some time to figure out what data type and file type available and what your datasets have.

`availTCGA` can return all information you need:

```{r}
availTCGA()
```

Note not all datasets have these property, `showTCGA` can help you to check it. It will return all data in TCGA, you can use following code in RStudio and search your data.

```{r, eval=FALSE}
View(showTCGA())
```

**OR you can use shiny app provided by `UCSCXenaTools` to search**.

Run shiny by:

```{r, eval=FALSE}
UCSCXenaTools::XenaShiny()
```

Download by shiny is under consideration, I am try learning more about how to operate shiny.


## SessionInfo

```{r}
sessionInfo()
```

## Bug Report

I have no time to test if all condition are right and all datasets can normally be downloaded. So if you have any question or suggestion, please open an issue on Github at <https://github.com/ShixiangWang/UCSCXenaTools/issues>. 

## Acknowledgement

This package is based on [XenaR](https://github.com/mtmorgan/XenaR), thanks [Martin Morgan](https://github.com/mtmorgan) for his work.

## LICENSE

GPL-3

please note, code from XenaR package under Apache 2.0 license.

## ToDo

* Shinny and more
